pkg/controllers/gracefuleviction/resource_binding_controller.go

```diff
@@ -0,0 +1,142 @@
+type ResourceBindingController struct {
+	client.Client
+	EventRecorder           record.EventRecorder
+	RateLimiterOptions      ratelimiterflag.Options
+	GracefulEvictionTimeout time.Duration
+}
+
+// Reconcile performs a full reconciliation for the object referred to by the Request.
+// The Controller will requeue the Request to be processed again if an error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (c *ResourceBindingController) Reconcile(ctx context.Context, req controllerruntime.Request) (controllerruntime.Result, error) {
+	klog.V(4).Infof("Reconciling ResourceBinding %s.", req.NamespacedName.String())
+
+	binding := &workv1alpha2.ResourceBinding{}
+	if err := c.Client.Get(context.TODO(), req.NamespacedName, binding); err != nil {
+		if apierrors.IsNotFound(err) {
+			return controllerruntime.Result{}, nil
+		}
+		return controllerruntime.Result{Requeue: true}, err
+	}
+
+	if !binding.DeletionTimestamp.IsZero() {
+		return controllerruntime.Result{}, nil
+	}
+
+	retryDuration, err := c.syncBinding(binding)
+	if err != nil {
+		return controllerruntime.Result{Requeue: true}, err
+	}
+	if retryDuration > 0 {
+		klog.V(4).Infof("Retry to evict task after %v minutes.", retryDuration.Minutes())
+		return controllerruntime.Result{RequeueAfter: retryDuration}, nil
+	}
+	return controllerruntime.Result{}, nil
+}
+
+func (c *ResourceBindingController) syncBinding(binding *workv1alpha2.ResourceBinding) (time.Duration, error) {
+	if len(binding.Spec.GracefulEvictionTasks) == 0 {
+		return 0, nil
+	}
+	processedTasks, retryDuration := processEvictionTasks(binding.Spec.GracefulEvictionTasks, c.GracefulEvictionTimeout)
+	if reflect.DeepEqual(binding.Spec.GracefulEvictionTasks, processedTasks) {
+		return retryDuration, nil
+	}
+
+	objPatch := client.MergeFrom(binding.DeepCopyObject().(client.Object))
+	binding.Spec.GracefulEvictionTasks = processedTasks
+	err := c.Client.Patch(context.TODO(), binding, objPatch)
+	if err != nil {
+		return 0, err
+	}
+
+	return retryDuration, nil
+}
+
+func processEvictionTasks(tasks []workv1alpha2.GracefulEvictionTask, timeout time.Duration) (
+	[]workv1alpha2.GracefulEvictionTask, time.Duration) {
+	var processedTasks []workv1alpha2.GracefulEvictionTask
+	var retryDuration time.Duration
+	var timeNow = time.Now()
+
+	retryDuration = timeout
+	for index := range tasks {
+		processedTask, taskRetryDuration := processEvictionTask(&tasks[index], timeNow, timeout)
+		if processedTask == nil {
+			continue
+		}
+
+		processedTasks = append(processedTasks, *processedTask)
+		if taskRetryDuration < retryDuration {
+			retryDuration = taskRetryDuration
+		}
+	}
+
+	// If the length of processed tasks is 0, we needn't retry in the feature.
+	if len(processedTasks) == 0 {
+		retryDuration = 0
+	}
+
+	return processedTasks, retryDuration
+}
+
+func processEvictionTask(evictionTask *workv1alpha2.GracefulEvictionTask, timeNow time.Time, timeout time.Duration) (
```

Comment:
When we remove it from both `spec.clusters` and `spec.gracefulEvictionTasks`, it will be clean by `removeOrphanWorks` in `pkg/controllers/binding/binding_controller.go:98`.

Reply:
Interesting, so when move the result from `spec.clusters` to `spec.gracefulEvictionTasks`, eviction does not actually begin. When we remove it from `spec.gracefulEvictionTasks`, the task is totally done and work is deleted, right?

Reply:
Yes, it it.

Reply:
Cool.